Analysis should not fail when highlighting has invalid text range

Parser can sometimes provide invalid text range for some symbols. We should not fail hard the analysis in such cases when saving highlighting. We should catch the exception and issue a warning. Warning should contain the information for the user that likely reason of the failure is incorrect source encoding and that it should be fixed by setting the correct encoding (ps TypeScript actually requires utf-8, however JS does not). We should only issue the encoding hint when we detect that encoding is different.

We should not catch the exception in case when we are raising the issue, since this would compromise the user experience (issue not being raised).

https://community.sonarsource.com/t/the-last-analysis-failed-ay1bfgmeze98x1oe3-da/108210/4

```
Caused by: java.lang.IllegalArgumentException: 6 is not a valid line offset for pointer. File src/common/video_conference/audio_video_settings/notary_alert.tsx has 0 character(s) at line 41
Process terminated with code: 1, time elapsed PT2M16.825569489S
5
Analysis failed:
6
7
at org.sonar.api.batch.fs.internal.DefaultInputFile.newPointer(DefaultInputFile.java:271)
8
at org.sonar.api.batch.fs.internal.DefaultInputFile.newRange(DefaultInputFile.java:287)
9
at org.sonar.plugins.javascript.bridge.BridgeServer$Location.toTextRange(BridgeServer.java:233)
10
at org.sonar.plugins.javascript.bridge.AnalysisProcessor.saveHighlights(AnalysisProcessor.java:200)
11
at org.sonar.plugins.javascript.bridge.AnalysisProcessor.processResponse(AnalysisProcessor.java:117)
12
at org.sonar.plugins.javascript.bridge.AnalysisWithProgram.analyze(AnalysisWithProgram.java:179)
13
at org.sonar.plugins.javascript.bridge.AnalysisWithProgram.analyzeProgram(AnalysisWithProgram.java:147)
14
at org.sonar.plugins.javascript.bridge.AnalysisWithProgram.analyzeFiles(AnalysisWithProgram.java:100)
15
at org.sonar.plugins.javascript.bridge.JsTsSensor.analyzeFiles(JsTsSensor.java:132)
16
at org.sonar.plugins.javascript.bridge.AbstractBridgeSensor.execute(AbstractBridgeSensor.java:79)
17
... 33 more
18
ERROR:
19
ERROR: Re-run SonarScanner using the -X switch to enable full debug logging.
20
at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:123)
21
at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:109)
22
at org.sonar.scanner.bootstrap.GlobalContainer.doAfterStart(GlobalContainer.java:125)
23
at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:123)
24
at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:109)
25
at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:57)
26
at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:51)
27
at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:46)
28
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
29
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
30
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
31
at java.base/java.lang.reflect.Method.invoke(Method.java:568)
32
at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
33
at jdk.proxy1/jdk.proxy1.$Proxy0.execute(Unknown Source)
34
at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:189)
35
at org.sonarsource.scanner.api.EmbeddedScanner.execute(EmbeddedScanner.java:138)
36
at org.sonarsource.scanner.cli.Main.execute(Main.java:126)
37
at org.sonarsource.scanner.cli.Main.execute(Main.java:81)
38
at org.sonarsource.scanner.cli.Main.main(Main.java:62)
39
Caused by: java.lang.IllegalArgumentException: 6 is not a valid line offset for pointer. File src/common/video_conference/audio_video_settings/notary_alert.tsx has 0 character(s) at line 41
40
at org.sonar.api.utils.Preconditions.checkArgument(Preconditions.java:43)
41
at org.sonar.api.batch.fs.internal.DefaultInputFile.checkValid(DefaultInputFile.java:338)
42
at org.sonar.scanner.sensor.ModuleSensorsExecutor.lambda$execute$1(ModuleSensorsExecutor.java:48)
43
at org.sonar.scanner.sensor.ModuleSensorsExecutor.withModuleStrategy(ModuleSensorsExecutor.java:66)
44
at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:48)
45
at org.son
```

--- a/sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/AnalysisProcessor.java
+++ b/sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/AnalysisProcessor.java
@@ -191,10 +191,15 @@ public class AnalysisProcessor {
   private void saveHighlights(BridgeServer.Highlight[] highlights) {
     NewHighlighting highlighting = context.newHighlighting().onFile(file);
     for (BridgeServer.Highlight highlight : highlights) {
-      highlighting.highlight(
-        highlight.location.toTextRange(file),
-        TypeOfText.valueOf(highlight.textType)
-      );
+      try {
+        highlighting.highlight(
+          highlight.location.toTextRange(file),
+          TypeOfText.valueOf(highlight.textType)
+        );
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Failed to save highlight", e);
+        // continue processing other highlights
+      }
     }
     highlighting.save();
   }
@@ -203,19 +208,24 @@ public class AnalysisProcessor {
     NewSymbolTable symbolTable = context.newSymbolTable().onFile(file);
     for (BridgeServer.HighlightedSymbol highlightedSymbol : highlightedSymbols) {
       BridgeServer.Location declaration = highlightedSymbol.declaration;
-      NewSymbol newSymbol = symbolTable.newSymbol(
-        declaration.startLine,
-        declaration.startCol,
-        declaration.endLine,
-        declaration.endCol
-      );
-      for (BridgeServer.Location reference : highlightedSymbol.references) {
-        newSymbol.newReference(
-          reference.startLine,
-          reference.startCol,
-          reference.endLine,
-          reference.endCol
+      try {
+        NewSymbol newSymbol = symbolTable.newSymbol(
+          declaration.startLine,
+          declaration.startCol,
+          declaration.endLine,
+          declaration.endCol
         );
+        for (BridgeServer.Location reference : highlightedSymbol.references) {
+          newSymbol.newReference(
+            reference.startLine,
+            reference.startCol,
+            reference.endLine,
+            reference.endCol
+          );
+        }
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Failed to create symbol", e);
+        // continue processing other symbols
       }
     }
     symbolTable.save();

--- a/sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/AnalysisProcessor.java
+++ b/sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/AnalysisProcessor.java
@@ -197,7 +197,8 @@ public class AnalysisProcessor {
           TypeOfText.valueOf(highlight.textType)
         );
       } catch (IllegalArgumentException e) {
-        LOG.warn("Failed to save highlight", e);
+        LOG.warn("Failed to save highlight in {} at {}", file.uri(), highlight.location);
+        LOG.warn("Exception cause", e);
         // continue processing other highlights
       }
     }
@@ -208,24 +209,30 @@ public class AnalysisProcessor {
     NewSymbolTable symbolTable = context.newSymbolTable().onFile(file);
     for (BridgeServer.HighlightedSymbol highlightedSymbol : highlightedSymbols) {
       BridgeServer.Location declaration = highlightedSymbol.declaration;
+      NewSymbol newSymbol;
       try {
-        NewSymbol newSymbol = symbolTable.newSymbol(
-          declaration.startLine,
-          declaration.startCol,
-          declaration.endLine,
-          declaration.endCol
-        );
-        for (BridgeServer.Location reference : highlightedSymbol.references) {
+        newSymbol =
+          symbolTable.newSymbol(
+            declaration.startLine,
+            declaration.startCol,
+            declaration.endLine,
+            declaration.endCol
+          );
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Failed to create symbol declaration in {} at {}", file.uri(), declaration);
+        continue;
+      }
+      for (BridgeServer.Location reference : highlightedSymbol.references) {
+        try {
           newSymbol.newReference(
             reference.startLine,
             reference.startCol,
             reference.endLine,
             reference.endCol
           );
+        } catch (IllegalArgumentException e) {
+          LOG.warn("Failed to create symbol reference in {} at {}", file.uri(), reference);
         }
-      } catch (IllegalArgumentException e) {
-        LOG.warn("Failed to create symbol", e);
-        // continue processing other symbols
       }
     }
     symbolTable.save();

--- a/sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/BridgeServer.java
+++ b/sonar-plugin/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/bridge/BridgeServer.java
@@ -195,6 +195,15 @@ public interface BridgeServer extends Startable {
     int endLine;
     int endCol;
 
+    public Location() {}
+
+    Location(int startLine, int startCol, int endLine, int endCol) {
+      this.startLine = startLine;
+      this.startCol = startCol;
+      this.endLine = endLine;
+      this.endCol = endCol;
+    }
+
     public int getStartLine() {
       return startLine;
     }
@@ -229,6 +238,11 @@ public interface BridgeServer extends Startable {
 
     TextRange toTextRange(InputFile inputFile) {
       return inputFile.newRange(this.startLine, this.startCol, this.endLine, this.endCol);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("%d:%d-%d:%d", startLine, startCol, endLine, endCol);
     }
   }
 

--- a/sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/bridge/AnalysisProcessorTest.java
+++ b/sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/bridge/AnalysisProcessorTest.java
@@ -35,14 +35,11 @@ class AnalysisProcessorTest {
       .build();
     var response = new BridgeServer.AnalysisResponse();
     var highlight = new BridgeServer.Highlight();
-    highlight.location = new BridgeServer.Location();
-    highlight.location.startLine = 1;
-    highlight.location.startCol = 2;
-    highlight.location.endLine = 1;
-    highlight.location.endCol = 1; // invalid range startCol > endCol
+    highlight.location = new BridgeServer.Location(1, 2, 1, 1); // invalid range startCol > endCol
     response.highlights = new BridgeServer.Highlight[] { highlight };
     processor.processResponse(context, mock(JsTsChecks.class), file, response);
-    assertThat(logTester.logs()).contains("Failed to save highlight");
+    assertThat(logTester.logs())
+      .contains("Failed to save highlight in " + file.uri() + " at 1:2-1:1");
   }
 
   @Test
@@ -57,13 +54,16 @@ class AnalysisProcessorTest {
       .build();
     var response = new BridgeServer.AnalysisResponse();
     var symbol = new BridgeServer.HighlightedSymbol();
-    symbol.declaration = new BridgeServer.Location();
-    symbol.declaration.startLine = 1;
-    symbol.declaration.startCol = 2;
-    symbol.declaration.endLine = 1;
-    symbol.declaration.endCol = 1; // invalid range startCol > endCol
+    symbol.declaration = new BridgeServer.Location(1, 2, 1, 1); // invalid range startCol > endCol
     response.highlightedSymbols = new BridgeServer.HighlightedSymbol[] { symbol };
     processor.processResponse(context, mock(JsTsChecks.class), file, response);
-    assertThat(logTester.logs()).contains("Failed to create symbol");
+    assertThat(logTester.logs())
+      .contains("Failed to create symbol declaration in " + file.uri() + " at 1:2-1:1");
+
+    symbol.declaration = new BridgeServer.Location(1, 1, 1, 2);
+    symbol.references = new BridgeServer.Location[] { new BridgeServer.Location(2, 2, 2, 1) };
+    processor.processResponse(context, mock(JsTsChecks.class), file, response);
+    assertThat(logTester.logs())
+      .contains("Failed to create symbol reference in " + file.uri() + " at 2:2-2:1");
   }
 }

--- a/sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/bridge/AnalysisProcessorTest.java
+++ b/sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/bridge/AnalysisProcessorTest.java
@@ -60,6 +60,7 @@ class AnalysisProcessorTest {
     assertThat(logTester.logs())
       .contains("Failed to create symbol declaration in " + file.uri() + " at 1:2-1:1");
 
+    context = SensorContextTester.create(baseDir);
     symbol.declaration = new BridgeServer.Location(1, 1, 1, 2);
     symbol.references = new BridgeServer.Location[] { new BridgeServer.Location(2, 2, 2, 1) };
     processor.processResponse(context, mock(JsTsChecks.class), file, response);

--- a/sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/bridge/AnalysisProcessorTest.java
+++ b/sonar-plugin/sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/bridge/AnalysisProcessorTest.java
@@ -1,0 +1,69 @@
+package org.sonar.plugins.javascript.bridge;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.nio.file.Path;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.sonar.api.batch.fs.internal.TestInputFileBuilder;
+import org.sonar.api.batch.sensor.internal.SensorContextTester;
+import org.sonar.api.issue.NoSonarFilter;
+import org.sonar.api.measures.FileLinesContext;
+import org.sonar.api.measures.FileLinesContextFactory;
+import org.sonar.api.testfixtures.log.LogTesterJUnit5;
+
+class AnalysisProcessorTest {
+
+  @TempDir
+  Path baseDir;
+
+  @org.junit.jupiter.api.extension.RegisterExtension
+  LogTesterJUnit5 logTester = new LogTesterJUnit5();
+
+  @Test
+  void should_not_fail_when_invalid_range() {
+    var fileLinesContextFactory = mock(FileLinesContextFactory.class);
+    when(fileLinesContextFactory.createFor(any())).thenReturn(mock(FileLinesContext.class));
+    var processor = new AnalysisProcessor(mock(NoSonarFilter.class), fileLinesContextFactory);
+    var context = SensorContextTester.create(baseDir);
+    var file = TestInputFileBuilder
+      .create("moduleKey", "file.js")
+      .setContents("var x  = 1;")
+      .build();
+    var response = new BridgeServer.AnalysisResponse();
+    var highlight = new BridgeServer.Highlight();
+    highlight.location = new BridgeServer.Location();
+    highlight.location.startLine = 1;
+    highlight.location.startCol = 2;
+    highlight.location.endLine = 1;
+    highlight.location.endCol = 1; // invalid range startCol > endCol
+    response.highlights = new BridgeServer.Highlight[] { highlight };
+    processor.processResponse(context, mock(JsTsChecks.class), file, response);
+    assertThat(logTester.logs()).contains("Failed to save highlight");
+  }
+
+  @Test
+  void should_not_fail_when_invalid_symbol() {
+    var fileLinesContextFactory = mock(FileLinesContextFactory.class);
+    when(fileLinesContextFactory.createFor(any())).thenReturn(mock(FileLinesContext.class));
+    var processor = new AnalysisProcessor(mock(NoSonarFilter.class), fileLinesContextFactory);
+    var context = SensorContextTester.create(baseDir);
+    var file = TestInputFileBuilder
+      .create("moduleKey", "file.js")
+      .setContents("var x  = 1;")
+      .build();
+    var response = new BridgeServer.AnalysisResponse();
+    var symbol = new BridgeServer.HighlightedSymbol();
+    symbol.declaration = new BridgeServer.Location();
+    symbol.declaration.startLine = 1;
+    symbol.declaration.startCol = 2;
+    symbol.declaration.endLine = 1;
+    symbol.declaration.endCol = 1; // invalid range startCol > endCol
+    response.highlightedSymbols = new BridgeServer.HighlightedSymbol[] { symbol };
+    processor.processResponse(context, mock(JsTsChecks.class), file, response);
+    assertThat(logTester.logs()).contains("Failed to create symbol");
+  }
+}

